### ISSUE TITLE
When lock is dropped, is data also dropped?

### ISSUE THREAD

**Issue by choejms** [kaist-cp/cs431#635]
Hi, I have a question on the lifetime of LockGuard, Lock and data.

It is my understanding that when LockGuard is dropped, Lock is also dropped because they have same lifetime annotation.
However, since data is a field of Lock, shouldn't data also be dropped?

How can other threads continue to use the data when Lock is dropped?
  
```
  pub struct Lock<L: RawLock, T> {
      lock: L,
      data: UnsafeCell<T>,
  }
```

**Comment by choejms** [comment]
Hi, I actually realized I had  a misconception.
When Lockguard is dropped, Lock is not dropped.
Instead lock is released (unlocked). 

But when lock is dropped, does the data not get dropped?

**Comment by kingdoctor123** [comment]
I think data will automatically be dropped, and it is natural design choice.

Unlike `Arc` which we clone multiple times to make duplicates, we usually pass immutable reference to the lock to each thread by wrapping `Lock` with `Arc`. (see https://doc.rust-lang.org/std/sync/struct.Mutex.html#examples) (see also https://users.rust-lang.org/t/why-mutex-doesnt-implement-clone-trait/34959 to check that Rust's `Mutex` does not implement `Clone` trait)

Since original lock does not get dropped when its reference gets out of scope, lock is dropped only once eventually, so we should drop its protected data at that moment too.

**Comment by Lee-Janggun** [comment]
Yes, the data is dropped.

These kind of questions can also be answered by implementing your won `Drop` trait, as in the [Rust book example](https://doc.rust-lang.org/book/ch15-03-drop.html).

Consider the following example using a Mutex:
```rust
use std::sync::Mutex;


#[derive(Default)]
struct S;

impl Drop for S {
    fn drop(&mut self) {
        println!("Dropping S");
    }
}

fn main() {
    let data = Mutex::new(S::default());
    
    drop(data);
    
    println!("Dropping data")
}
```

If dropping the mutex `data` drops the inner `S`, we would expect "Dropping S" to be printed before "Dropping Data". Indeed, running the example leads to the following result:
```
Dropping S
Dropping data
```
You can see it for yourself in this [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=41fa7294ff2ab83321c38b7b92194f80).

To add a bit more explanation to the original question, when we `drop` a `Lock`, it will recursively call the `drop` function for all of its field, as is defined in the [destructors reference](https://doc.rust-lang.org/reference/destructors.html#destructors).

Intuitively, this make sense because since `Lock<T>` has `T` (and not a reference to `T`) as one of its field, having ownership of `Lock<T>` means you ownership for `T` as well, so giving up ownership of `Lock<T>` would mean giving up the ownership for `T`. The locking protocol is a way to share this ownership.

**Comment by choejms** [comment]
Thank you for the detailed explanation. Appreciate it!!
